File: sdk/go/worker.go
package steppr

import (
	"encoding/json"
	"fmt"
	"net/http"
	"runtime/debug"
	"sort"
	"sync"
)

// HandlerFunc is the signature a node handler implements. It receives the
// inbound WorkflowMessage and returns either a *NodeResponse or
// *FilterResponse. Returning a non-nil error converts to a "failure"
// NodeResponse with code HANDLER_ERROR; structured failures should be
// returned as a NodeResponse with Status="failure" and a populated Error.
type HandlerFunc func(msg *WorkflowMessage) (any, error)

// Worker is a Steppr HTTP worker. Register handlers per node ID with
// Handler, then call Run to start serving on the configured port.
type Worker struct {
	Port int

	mu       sync.RWMutex
	handlers map[string]HandlerFunc
}

// New returns a Worker that will listen on the given port when Run is
// called. The zero value of Worker is also usable; only Port matters.
func New(port int) *Worker {
	return &Worker{Port: port, handlers: map[string]HandlerFunc{}}
}

// Handler registers a handler for the given node ID. Re-registering
// overwrites the prior handler.
func (w *Worker) Handler(nodeID string, fn HandlerFunc) {
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.handlers == nil {
		w.handlers = map[string]HandlerFunc{}
	}
	w.handlers[nodeID] = fn
}

// HandlerNode is a convenience for the common case where the handler
// always returns a NodeResponse (Enhancer or Action). The closure
// receives the message and returns the output payload to merge.
//   w.HandlerNode("validate_order", func(msg *WorkflowMessage) (map[string]any, error) {
//       return map[string]any{"validated": true}, nil
//   })
func (w *Worker) HandlerNode(nodeID string, fn func(*WorkflowMessage) (map[string]any, error)) {
	w.Handler(nodeID, func(msg *WorkflowMessage) (any, error) {
		out, err := fn(msg)
		if err != nil {
			return nil, err
		}
		return &NodeResponse{
			ExecutionID: msg.ExecutionID,
			NodeID:      msg.NodeID,
			Status:      "success",
			Output:      out,
		}, nil
	})
}

// Run starts the HTTP server. It blocks until the server returns an error.
func (w *Worker) Run() error {
	mux := http.NewServeMux()
	mux.HandleFunc("/handlers", w.serveHandlers)
	mux.HandleFunc("/execute", w.serveExecute)

	addr := fmt.Sprintf(":%d", w.Port)
	fmt.Printf("steppr worker listening on %s (%d handler(s))\n", addr, len(w.handlers))
	return http.ListenAndServe(addr, mux)
}

func (w *Worker) serveHandlers(rw http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(rw, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	w.mu.RLock()
	ids := make([]string, 0, len(w.handlers))
	for id := range w.handlers {
		ids = append(ids, id)
	}
	w.mu.RUnlock()
	sort.Strings(ids)
	writeJSON(rw, http.StatusOK, map[string]any{"node_ids": ids})
}

func (w *Worker) serveExecute(rw http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(rw, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var msg WorkflowMessage
	if err := json.NewDecoder(r.Body).Decode(&msg); err != nil {
		writeJSON(rw, http.StatusBadRequest, map[string]any{
			"error": fmt.Sprintf("decoding request: %v", err),
		})
		return
	}

	w.mu.RLock()
	fn := w.handlers[msg.NodeID]
	w.mu.RUnlock()
	if fn == nil {
		writeJSON(rw, http.StatusBadRequest, map[string]any{
			"error": fmt.Sprintf("no handler for node %q", msg.NodeID),
		})
		return
	}

	result, err := safeInvoke(fn, &msg)
	if err != nil {
		writeJSON(rw, http.StatusInternalServerError, &NodeResponse{
			ExecutionID: msg.ExecutionID,
			NodeID:      msg.NodeID,
			Status:      "failure",
			Output:      map[string]any{},
			Error:       &ErrorInfo{Code: "HANDLER_ERROR", Message: err.Error()},
		})
		return
	}

	writeJSON(rw, http.StatusOK, result)
}

// safeInvoke runs fn and converts panics into normal errors so a worker
// never crashes the process on a buggy handler.
func safeInvoke(fn HandlerFunc, msg *WorkflowMessage) (result any, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("handler panic: %v\n%s", r, debug.Stack())
		}
	}()
	return fn(msg)
}

func writeJSON(rw http.ResponseWriter, status int, body any) {
	data, err := json.Marshal(body)
	if err != nil {
		http.Error(rw, "marshal error", http.StatusInternalServerError)
		return
	}
	rw.Header().Set("Content-Type", "application/json")
	rw.WriteHeader(status)
	_, _ = rw.Write(data)
}

File: sdk/typescript/src/worker.ts
/**
 * Node-only StepprWorker. Wraps a handler registry behind an HTTP server
 * exposing GET /handlers (discovery) and POST /execute (invocation).
 *
 * Mirrors the Python SDK API. Browser code should import from "@steppr/sdk"
 * (the core package) and use InMemoryResolver instead.
 */

import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import type { NodeResponse, WorkflowMessage } from "./types.js";

export type Handler = (msg: WorkflowMessage) =>
  | NodeResponse
  | Promise<NodeResponse>;

export interface StepprWorkerOptions {
  /** Port to listen on. Default 8080. */
  port?: number;
  /** Address to bind. Default 0.0.0.0. */
  host?: string;
}

export class StepprWorker {
  private readonly handlers = new Map<string, Handler>();
  private readonly opts: Required<StepprWorkerOptions>;

  constructor(opts: StepprWorkerOptions = {}) {
    this.opts = {
      port: opts.port ?? 8080,
      host: opts.host ?? "0.0.0.0",
    };
  }

  /** Register a handler for a node ID. */
  handler(nodeId: string, fn: Handler): this {
    this.handlers.set(nodeId, fn);
    return this;
  }

  /** Start the HTTP server. Resolves to a stop function. */
  async run(): Promise<() => Promise<void>> {
    const server = createServer((req, res) => this.dispatch(req, res));
    await new Promise<void>((resolve) => {
      server.listen(this.opts.port, this.opts.host, resolve);
    });
    // eslint-disable-next-line no-console
    console.log(
      `steppr worker listening on :${this.opts.port} (${this.handlers.size} handler(s))`,
    );
    return () =>
      new Promise<void>((resolve, reject) => {
        server.close((err) => (err ? reject(err) : resolve()));
      });
  }

  private async dispatch(req: IncomingMessage, res: ServerResponse): Promise<void> {
    if (req.method === "GET" && req.url === "/handlers") {
      respond(res, 200, {
        node_ids: [...this.handlers.keys()].sort(),
      });
      return;
    }
    if (req.method === "POST" && req.url === "/execute") {
      let body: WorkflowMessage;
      try {
        body = await readJson<WorkflowMessage>(req);
      } catch (err) {
        respond(res, 400, {
          error: err instanceof Error ? err.message : String(err),
        });
        return;
      }
      const fn = this.handlers.get(body.node_id);
      if (!fn) {
        respond(res, 400, { error: `no handler for node ${body.node_id}` });
        return;
      }
      try {
        const result = await fn(body);
        respond(res, 200, result);
      } catch (err) {
        respond(res, 500, {
          execution_id: body.execution_id,
          node_id: body.node_id,
          status: "failure",
          error: {
            code: "HANDLER_ERROR",
            message: err instanceof Error ? err.stack ?? err.message : String(err),
          },
        });
      }
      return;
    }
    respond(res, 404, { error: "not found" });
  }
}

function respond(res: ServerResponse, status: number, body: unknown): void {
  const data = JSON.stringify(body);
  res.statusCode = status;
  res.setHeader("content-type", "application/json");
  res.setHeader("content-length", String(Buffer.byteLength(data)));
  res.end(data);
}

async function readJson<T>(req: IncomingMessage): Promise<T> {
  const chunks: Buffer[] = [];
  for await (const chunk of req) {
    chunks.push(chunk as Buffer);
  }
  const text = Buffer.concat(chunks).toString("utf-8");
  if (!text) return {} as T;
  return JSON.parse(text) as T;
}
